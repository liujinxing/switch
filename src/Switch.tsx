import React from 'react';
import styled from 'styled-components';
import classNames from 'classnames';
import SwitchTrack from './SwitchTrack';
import SwitchBase from './SwitchBase';

interface SwitchProps {
  color?: string;
  disabled?: boolean;
  checked?: boolean;
  value: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const Wrapper = styled.span<{ disabled?: boolean }>`
  width: 58px;
  height: 38px;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  flex-shrink: 0;
  vertical-align: middle;
  cursor: ${(props) => (props.disabled ? 'default' : 'pointer')};
`;

export default function Switch(props: SwitchProps) {
  const { disabled } = props;
  return (
    <Wrapper className={classNames('sinoui-switch')} disabled={disabled}>
      <SwitchBase {...props} />
      <SwitchTrack className="sinoui-switch-track" />
    </Wrapper>
  );
}
