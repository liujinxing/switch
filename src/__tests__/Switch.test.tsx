import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Switch from 'src/Switch';
import TestWrapper from './TestWrapper';

describe('Switch 镜像测试', () => {
  it('渲染Switch', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Switch value="switch1" />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('不可用状态', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Switch value="switch2" disabled />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('选中状态', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Switch value="switch3" checked color="accent" />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
