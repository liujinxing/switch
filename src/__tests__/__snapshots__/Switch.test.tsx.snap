// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch 镜像测试 不可用状态 1`] = `
.c2 {
  width: 100%;
  height: 100%;
  opacity: 0.38;
  z-index: -1;
  -webkit-transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  border-radius: 7px;
  background-color: #000000;
  opacity: 0.12;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  position: absolute;
  top: 0;
  left: 0;
  color: #fafafa;
  z-index: 1;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  padding: 9px;
  border-radius: 50%;
  color: #bdbdbd;
}

.c1:hover {
  background-color: rgba(0,0,0,0.08);
}

.c1:hover {
  background-color: transparent;
}

.c1 .sinoui-switch__input {
  top: 0;
  left: -100%;
  width: 300%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  opacity: 0;
  padding: 0;
  z-index: 1;
  position: absolute;
}

.c1 .sinoui-switch__thumb {
  width: 20px;
  height: 20px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  border-radius: 50%;
  background-color: currentColor;
}

.c1 .sinoui-switch__ripple {
  width: 38px;
  height: 38px;
}

.c0 {
  width: 58px;
  height: 38px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  cursor: default;
}

<span
  className="c0 sinoui-switch sinoui-switch--disabled"
  disabled={true}
>
  <span
    className="c1 sinoui-switch__label"
    disabled={true}
  >
    <input
      className="sinoui-switch__input"
      disabled={true}
      type="checkbox"
      value="switch2"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
    className="c2 sinoui-switch__track"
    disabled={true}
  />
</span>
`;

exports[`Switch 镜像测试 渲染Switch 1`] = `
.c2 {
  width: 100%;
  height: 100%;
  opacity: 0.38;
  z-index: -1;
  -webkit-transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  border-radius: 7px;
  background-color: #000000;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  position: absolute;
  top: 0;
  left: 0;
  color: #fafafa;
  z-index: 1;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  padding: 9px;
  border-radius: 50%;
}

.c1:hover {
  background-color: rgba(0,0,0,0.08);
}

.c1 .sinoui-switch__input {
  top: 0;
  left: -100%;
  width: 300%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  opacity: 0;
  padding: 0;
  z-index: 1;
  position: absolute;
}

.c1 .sinoui-switch__thumb {
  width: 20px;
  height: 20px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  border-radius: 50%;
  background-color: currentColor;
}

.c1 .sinoui-switch__ripple {
  width: 38px;
  height: 38px;
}

.c0 {
  width: 58px;
  height: 38px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  cursor: pointer;
}

<span
  className="c0 sinoui-switch"
>
  <span
    className="c1 sinoui-switch__label"
  >
    <input
      className="sinoui-switch__input"
      type="checkbox"
      value="switch1"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
    className="c2 sinoui-switch__track"
  />
</span>
`;

exports[`Switch 镜像测试 选中状态 1`] = `
.c2 {
  width: 100%;
  height: 100%;
  opacity: 0.38;
  z-index: -1;
  -webkit-transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: opacity 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,background-color 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  border-radius: 7px;
  background-color: #000000;
  background-color: #ff4081;
  opacity: 0.5;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  position: absolute;
  top: 0;
  left: 0;
  color: #fafafa;
  z-index: 1;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: left 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 150ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  padding: 9px;
  border-radius: 50%;
  -webkit-transform: translateX(50%);
  -ms-transform: translateX(50%);
  transform: translateX(50%);
  color: #ff4081;
}

.c1:hover {
  background-color: rgba(0,0,0,0.08);
}

.c1:hover {
  background-color: rgba(255,64,129,0.12);
}

.c1 .sinoui-switch__input {
  top: 0;
  left: -100%;
  width: 300%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  opacity: 0;
  padding: 0;
  z-index: 1;
  position: absolute;
}

.c1 .sinoui-switch__thumb {
  width: 20px;
  height: 20px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  border-radius: 50%;
  background-color: currentColor;
}

.c1 .sinoui-switch__ripple {
  width: 38px;
  height: 38px;
}

.c0 {
  width: 58px;
  height: 38px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  cursor: pointer;
}

<span
  className="c0 sinoui-switch sinoui-switch--checked"
>
  <span
    checked={true}
    className="c1 sinoui-switch__label"
    color="accent"
  >
    <input
      checked={true}
      className="sinoui-switch__input"
      type="checkbox"
      value="switch3"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
    checked={true}
    className="c2 sinoui-switch__track"
    color="accent"
  />
</span>
`;
